src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/ComplexTypeModelBinder.cs

```diff
@@ -357,24 +358,32 @@ protected virtual object CreateModel(ModelBindingContext bindingContext)
             // application developer should know that this was an invalid type to try to bind to.
             if (_modelCreator == null)
             {
-                // The following check causes the ComplexTypeModelBinder to NOT participate in binding structs as 
+                // The following check causes the ComplexTypeModelBinder to NOT participate in binding structs as
                 // reflection does not provide information about the implicit parameterless constructor for a struct.
                 // This binder would eventually fail to construct an instance of the struct as the Linq's NewExpression
                 // compile fails to construct it.
                 var modelTypeInfo = bindingContext.ModelType.GetTypeInfo();
                 if (modelTypeInfo.IsAbstract || modelTypeInfo.GetConstructor(Type.EmptyTypes) == null)
                 {
-                    if (bindingContext.IsTopLevelObject)
```

Comment:
Curious, why was this check removed?

Reply:
This was deciding between two messages and it did so incorrectly. In the `!IsTopLevelObject` case, `bindingContext.ModelMetadata.ContainerType.FullName` could throw for collection elements before we added parameter metadata. Now it'll throw for type metadata as well.